static graph api skeleton

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54995 static graph api skeleton**
* #54981 refactor autograd_hook
* #56730 add a test for grad view with torch amp

provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

zhaojuanmao committed:
static graph api skeleton

provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1559,6 +1559,14 @@ void Reducer::record_backward_comm_end_time() {
 }
 
+void Reducer::set_static_graph() {
+  TORCH_INTERNAL_ASSERT(
```
nit: This should be `TORCH_CHECK` since it is a user error and not an internal error.

pritamdamania87 commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -122,6 +122,9 @@ class Reducer {
   // recorded once every "sample_rate" training iterations.
   void set_ddp_runtime_logging_sample_rate(int sample_rate);
 
+  // Speficy the training graph is static.
```
nit: Specify

pritamdamania87 commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -416,6 +419,8 @@ class Reducer {
   // Division factor for reduction of gradients.
   int divFactor_;
 
+  bool static_graph_ = false;
```
nit: initialize this to false in the constructor initializer instead.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -1385,3 +1385,22 @@ def set_ddp_runtime_logging_sample_rate(self, sample_rate):
             )
         self.reducer._set_ddp_runtime_logging_sample_rate(sample_rate)
+
+    # Users can explicitly let DDP know the trained graph is static,
+    # when 1) the set of used and unused parameters will not change
+    # during the whole training loop; in this case, it does not matter
+    # whether users set find_unsued_parameters = true or not.
+    # 2) how the graph is trained will not change during the whole training
+    # loop (meaning there is no control flow depending on iterations).
+    # When graph is set to be static, DDP will support cases that can not
+    # be supported in the past: 1) reentrant backwards
+    # 2) activation checkpointing multiple times 3)
+    # activation checkpointing with find_unused_parameters = true.
+    # 4) not all output tensors are used in loss calculation.
+    # 5) there is model parameter that is outside of forward function.
+    # 6) potentially improve performance when find_unsued_parameters = true
+    # or there are unused parameters, as DDP will not search graph in each
+    # iteraton to detect unused parameters when static_graph is set to be True.
```
Why aren't we using regular python docs for this method?

zhaojuanmao replied:
will do

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -90,6 +90,10 @@ void Logger::set_comm_hook(const std::string& hook) {
 }
 
+void Logger::set_static_graph() {
+  ddp_logging_data_->static_graph = reducer_->static_graph_;
```
do we need lock here?

zhaojuanmao replied:
yes, good catch, thanks!

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3705,6 +3705,28 @@ def test_ddp_logging_data_gpu(self):
+                    loss,
+                    1,
+                )
+                model_DDP._set_static_graph()
```
What does setting static graph at the end test?

zhaojuanmao replied:
it tests whether errors are threw if we set_static_graph() after training starts

rohan-varma reviewed (approved):
Looks good, but can we also add a test that the logger correctly picks it up as a static graph? We can add that as part of the existing logger tests. Also would be good to address @pritamdamania 's comments. Thanks!

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1385,3 +1385,22 @@ def set_ddp_runtime_logging_sample_rate(self, sample_rate):
+    # 6) potentially improve performance when find_unsued_parameters = true
+    # or there are unused parameters, as DDP will not search graph in each
+    # iteraton to detect unused parameters when static_graph is set to be True.
+    def _set_static_graph(self):
```
I'm assuming that this can't be set after some training iterations have started (i.e. in the middle of training). Might be good to clarify that in the docs (needs to be set before any training, and needs to be set across all ranks) to avoid confusion

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "static graph api skeleton"


provide an DDP private API to explicitly set the training is static, also set this flag in logger

Differential Revision: [D27444965](https://our.internmc.facebook.com/intern/diff/D27444965/)

[ghstack-poisoned]
